### PR TITLE
[Script Execution] Script execution module

### PR DESCRIPTION
**Pull request by sideninja**
Closes: #4640 

This PR adds a script execution module. The implementation creates a new module that configures FVM for script execution and [depends on the `RegistersAtHeight` function](https://github.com/onflow/flow-go/blob/d3e870bd883ee84762c6bd6953b9e1bf6775b973/module/execution/scripts.go#L24) to fetch values when it needs to read them. That function is then provided by the indexer module which is required for script execution to function.

### CODE REVIEW

File: admin/commands/state_synchronization/execute_script.go
Comment (sideninja) [inline admin/commands/state_synchronization/execute_script.go:22]: I'm fine with removing this command, but before we have #4781 I need it to test.

File: cmd/access/node_builder/access_node_builder.go
Comment (sideninja) [inline cmd/access/node_builder/access_node_builder.go:746]: I'm not sure if this will get initialized only after the ingestion is done cc @peterargue

File: admin/commands/state_synchronization/execute_script.go
Comment (koko1123) [reply admin/commands/state_synchronization/execute_script.go:22]: I think we can keep this command even after #4781 is merged

File: cmd/access/node_builder/access_node_builder.go
Comment (peterargue) [reply cmd/access/node_builder/access_node_builder.go:749]: no, since the indexers is a `DependableComponent`, anything that depends on it must be too. This doesn't really look like a component either since it doesn't implement the `ReadyDoneAware` interface. You could probably just run this code at the end of the indexer closure, and we can figure out a better place for it later when integrating it into the access api

File: module/execution/scripts.go
Comment (zhangchiqing) [inline module/execution/scripts.go:107]: we only get one register at a time, do we need `registersAtHeight` or just `registerAtHeight`?

Comment (sideninja) [reply module/execution/scripts.go:107]: I just thought I would make the function compatible with the indexer existing function, I can make an adapter and make this function only take a single register. I personally don't mind either way. One adds a bit more code the other is not as perfect since we always just need one like you said.

File: module/execution/scripts.go
Comment (peterargue) [inline module/execution/scripts.go:124]:
```suggestion
		if len(values) != 1 {
```

**Review by peterargue — APPROVED** [review]
looks good

File: cmd/access/node_builder/access_node_builder.go
Comment (zhangchiqing) [inline cmd/access/node_builder/access_node_builder.go:755]: I wonder why `ExecutionIndexCore` provides a `RegisterValues` method which get a list of registers?

I think we can change the IndexCore's `RegisterValues` method to `RegisterValue` method, which only work for a single register, this way, we can also get rid of the IndexRegisterAdapter.

The reason I suggest this is that I don't see a use case where we need to get a list of registers at once. If that's the case, we'd better just keep things simple.

Comment (peterargue) [reply cmd/access/node_builder/access_node_builder.go:755]: I think the batch lookups will be needed in the `GetRegisters` endpoint.

Comment (zhangchiqing) [reply cmd/access/node_builder/access_node_builder.go:755]: Yeah, `GetRegisters` is not being used except some tests. I wonder if we can get rid of it?

File: module/execution/scripts.go
Comment (sideninja) [reply module/execution/scripts.go:124]: haha brainfart

File: cmd/access/node_builder/access_node_builder.go
Comment (sideninja) [reply cmd/access/node_builder/access_node_builder.go:755]: I don't know of a case where we would need a list beside that API. I looked into emulator implementation and it's also fetching a single register at a time. Since FVM always requests a single register I don't know what a usecase is for getting them in batch so I would think changing to a single register is fine, but with this we will break API compatibility. If that's acceptable then we can proceed. Or we could still have a loop that iterates the requested values and gets them.

Comment (zhangchiqing) [reply cmd/access/node_builder/access_node_builder.go:755]: I think we are talking about two things here: 
1. `ExecutionIndexerCore` has a `RegisterValues` which is not being used by anyone
2. `ledger` has a `GetRegisters` which is also not bing used within flow-go codebase, but not sure if any external repo depend on this.


Since `ExecutionIndexerCore` is created by us recently, and its `RegisterValues` is not being used by anyone, I'm suggesting to replace `RegisterValues` with `RegisterValue`, and get rid of the `IndexRegisterAdapter`

Comment (peterargue) [reply cmd/access/node_builder/access_node_builder.go:755]: that sounds reasonable to me. and we can easily add a plural version in the future if needed

Comment (koko1123) [reply cmd/access/node_builder/access_node_builder.go:755]: I think the `GetRegisters` in `ledger` is meant as an abstraction over `[]*ledger.Mtrie`, right? We will still need to implement the plural version for `ExecutionData.GetRegisters`

Comment (zhangchiqing) [reply cmd/access/node_builder/access_node_builder.go:755]: The `GetRegsiters` takes a `state commitment`, which means it's only for a single MTrie
https://github.com/onflow/flow-go/blob/master/storage/ledger.go#L14

Comment (peterargue) [reply cmd/access/node_builder/access_node_builder.go:755]: we are adding a new endpoint to the `ExecutionDataAPI` to get a set of register values (https://github.com/onflow/flow/pull/1389), which will need a method to get a list of registers for a given block height. However, I think that's separate from this discussion, since we could just as easily have 2 methods, `RegisterValue` and `RegisterValues`, or move the loop into the API implementation.

In this case, I think it makes the sense to optimize for the most common case which will be requesting a single register value at a time via script execution.

Comment (koko1123) [reply cmd/access/node_builder/access_node_builder.go:755]: +1 I think it's easier to move the loop to the API

Comment (zhangchiqing) [reply cmd/access/node_builder/access_node_builder.go:755]: I will approve for now. Lets add a RegisterValue to the `ExecutionIndexerCore` in a separate PR, and remove this adapter, since it's adding unnecessary layer.

**Review by zhangchiqing — APPROVED** [review]

**Review by koko1123 — APPROVED** [review]
Approving with intent to remove `ExecutionIndexerCore.RegisterValues` and `IndexRegisterAdapter` and replace with a `GetRegisters` implementation as part of https://github.com/onflow/flow-go/issues/4782

**Pull request merged**
